### PR TITLE
HierarchicalSampler

### PR DESCRIPTION
**Pull request by elias-ramzi**
Added HierarchicalSampler. It is the sampling scheme used in the FastAP paper.

Implementation adapted from : https://github.com/kunhe/Deep-Metric-Learning-Baselines/blob/master/datasets.py

For the first version, I made the HierarchicalSampler a batch sampler because it's closer to the original implementation.

Discussed in : https://github.com/KevinMusgrave/pytorch-metric-learning/issues/105.

### CODE REVIEW

**Comment by KevinMusgrave** [comment]
Thanks for doing this! I'll review the code in a few days.

**Comment by elias-ramzi** [comment]
Thanks for the corrections and the unittest.

I also had a question that is not specific to this PR : why do you set `pin_memory` to false by default for the DataLoader's ?

**Comment by KevinMusgrave** [comment]
> I also had a question that is not specific to this PR : why do you set `pin_memory` to false by default for the DataLoader's ?

I'm not sure actually. I think I was experimenting with that a long time ago when I was facing hanging issues with the dataloader. I could set it to True instead, if that speeds things up. I wonder why PyTorch doesn't set it to True by default

**Comment by elias-ramzi** [comment]
I think it's because you shouldn't use `pin_memory=True` when you are working on cpu's. But it might be more complicated than that ! 

Thanks for your help on this PR.
I think I will also do one on powerful-benchmarker for the SOP dataset to include the super classes.
